src/mappings/StakingRewards/shared.ts

```diff
@@ -0,0 +1,234 @@
+import {
+  getOrCreateStakingRewardsContract,
+  getTotalStakingRewards,
+  getTotalPlatformRewards,
+  isWithPlatformToken,
+} from '../../models/StakingRewardsContract'
+import { StakingRewards } from '../../../generated/templates/StakingRewards/StakingRewards'
+import { StakingRewardsWithPlatformToken } from '../../../generated/templates/StakingRewardsWithPlatformToken/StakingRewardsWithPlatformToken'
+import { getOrCreateRewardsVault } from '../../models/RewardsVault'
+
+import { RewardsVault as RewardsVaultTemplate } from '../../../generated/templates'
+
+function getContractForType<C>(
+  contractAddress: Address,
+  type: StakingRewardsContractType,
+): C {
+  return type === StakingRewardsContractType.STAKING_REWARDS_WITH_PLATFORM_TOKEN
+    ? (StakingRewardsWithPlatformToken.bind(contractAddress) as C)
+    : (StakingRewards.bind(contractAddress) as C)
+}
+
+export function handleRewardAddedForType<C>(
+  contractAddress: Address,
+  type: StakingRewardsContractType,
+): void {
+  let contract = getContractForType<C>(contractAddress, type)
+
+  let stakingRewards = getOrCreateStakingRewardsContract(
+    contractAddress,
+    StakingRewardsContractType.STAKING_REWARDS,
+  )
+
+  stakingRewards.lastUpdateTime = contract.lastUpdateTime().toI32()
+  stakingRewards.periodFinish = contract.periodFinish().toI32()
+  stakingRewards.rewardRate = contract.rewardRate()
+  stakingRewards.totalStakingRewards = getTotalStakingRewards(stakingRewards)
+
+  if (isWithPlatformToken(type)) {
+    stakingRewards.platformRewardRate = (contract as StakingRewardsWithPlatformToken).platformRewardRate()
+    stakingRewards.totalPlatformRewards = getTotalPlatformRewards(
+      stakingRewards,
+    )
+  }
+
+  stakingRewards.save()
+}
+
+export function handleStakedForType<C>(
+  contractAddress: Address,
+  type: StakingRewardsContractType,
+  user: Address,
+): void {
+  let contract = getContractForType<C>(contractAddress, type)
+
+  {
+    let stakingRewards = getOrCreateStakingRewardsContract(
+      contractAddress,
+      type,
+    )
+
+    stakingRewards.lastUpdateTime = contract.lastUpdateTime().toI32()
+    stakingRewards.rewardPerTokenStored = contract.rewardPerTokenStored()
+    stakingRewards.periodFinish = contract.periodFinish().toI32()
+    stakingRewards.rewardRate = contract.rewardRate()
+    stakingRewards.totalStakingRewards = getTotalStakingRewards(stakingRewards)
+
+    if (isWithPlatformToken(type)) {
+      stakingRewards.platformRewardPerTokenStored = (contract as StakingRewardsWithPlatformToken).platformRewardPerTokenStored()
+      stakingRewards.totalPlatformRewards = getTotalPlatformRewards(
+        stakingRewards,
+      )
+    }
+
+    stakingRewards.save()
+  }
+
+  {
+    let reward = getOrCreateStakingReward(
+      contractAddress,
+      user,
+      StakingRewardType.REWARD_PER_TOKEN_PAID,
+    )
+    reward.amount = contract.userRewardPerTokenPaid(user)
```

Comment:
Also need `contract.rewards(user)` <- this is the accumulated rewards

Reply:
Also add this to all `withdraw`, `stake`, etc

Reply:
Taken care of :+1: